pkg/protohttp/protohttp.go

```diff
@@ -131,24 +143,77 @@ func deserializePayloadFromReader(reader *bufio.Reader) ([]byte, error) {
 	return messageContentsAsBytes, nil
 }
 
-func checkIfResponseHasError(rsp *http.Response) error {
+// CheckIfResponseHasError checks an HTTP Response for errors and returns error
+// information with the following precedence:
+// 1. "linkerd-error" header, with protobuf-encoded apiError
+// 2. non-200 Status Code, with Kubernetes StatusError
+// 3. non-200 Status Code
+func CheckIfResponseHasError(rsp *http.Response) error {
+	// check for protobuf-encoded error
 	errorMsg := rsp.Header.Get(errorHeader)
-
 	if errorMsg != "" {
 		reader := bufio.NewReader(rsp.Body)
 		var apiError pb.ApiError
 
-		err := fromByteStreamToProtocolBuffers(reader, &apiError)
+		err := FromByteStreamToProtocolBuffers(reader, &apiError)
 		if err != nil {
 			return fmt.Errorf("Response has %s header [%s], but response body didn't contain protobuf error: %v", errorHeader, errorMsg, err)
 		}
 
 		return errors.New(apiError.Error)
 	}
 
+	// check for JSON-encoded error
 	if rsp.StatusCode != http.StatusOK {
+		if rsp.Body != nil {
+			bytes, err := ioutil.ReadAll(rsp.Body)
+			if err == nil && len(bytes) > 0 {
+				body := string(bytes)
+				obj, err := k8s.ToRuntimeObject(body)
+				if err == nil {
+					return fmt.Errorf("Unexpected API response: %s (%s)", rsp.Status, kerrors.FromObject(obj))
+				}
+				return fmt.Errorf("Unexpected API response: %s (%s)", rsp.Status, body)
+			}
+		}
+
 		return fmt.Errorf("Unexpected API response: %s", rsp.Status)
 	}
 
 	return nil
 }
+
+// FromByteStreamToProtocolBuffers converts a byte stream to a protobuf message.
+func FromByteStreamToProtocolBuffers(byteStreamContainingMessage *bufio.Reader, out proto.Message) error {
+	messageAsBytes, err := deserializePayloadFromReader(byteStreamContainingMessage)
+	if err != nil {
+		return fmt.Errorf("error reading byte stream header: %v", err)
+	}
+
+	err = proto.Unmarshal(messageAsBytes, out)
+	if err != nil {
+		return fmt.Errorf("error unmarshalling array of [%d] bytes error: %v", len(messageAsBytes), err)
+	}
+
+	return nil
+}
+
+// TapReqToURL converts a TapByResourceRequest protobuf object to a URL for use
+// with the Kubernetes tap.linkerd.io APIService.
+func TapReqToURL(req *pb.TapByResourceRequest) string {
```

Comment:
This feels pretty specific to the tap apiservice and a bit out of place in this file.

Reply:
I agree, but ended up putting it here for a few reasons:
- both `/cli/cmd` and `/controller/tap` depend on it, so it should be somewhere in `/pkg`
- `protohttp` already depends on `/controller/gen/public`, so putting `TapReqToURL` here does not add any new edges to our dependency tree
- it relates to the business of protobuf <-> http conversion

Given all this, the only other option would be to put that one function into a new package under `/pkg`, and that new package would still depend on `/controller/gen/public` (which is a more general issue to be fixed in #2751). Once #2751 completes, I'd be more open to a reorg of this code.